docs/website/contents/for-developers/utxo-hd/utxo-hd.md

```diff
@@ -0,0 +1,302 @@
+
+### Shelley instantiation and ledger tables
+
+The `LedgerState (ShelleyBlock proto era)` data family instances are augmented
+with a new field which will hold these entries that will be extracted from and
+injected to the `NewEpochState` before calling the Ledger rules. This new field
+(which we call _ledger tables_) is a container-like structure parametrized by a
+`Key` and `Value` type families.
+
+```diff haskell
+data instance LedgerState (ShelleyBlock proto era) mk = ShelleyLedgerState {
+       shelleyLedgerTip        :: !(WithOrigin (ShelleyTip proto era))
+     , shelleyLedgerState      :: !(SL.NewEpochState era)
+     , shelleyLedgerTransition :: !ShelleyTransition
++    , shelleyLedgerTables     :: !(LedgerTables (LedgerState (ShelleyBlock proto era)) mk)
+    }
+
+data LedgerTables l mk = LedgerTables {
+    getLedgerTables :: mk (Key l) (Value l)
+}
+```
+
+For a Shelley block, these type families are mapped to the same types as above:
+
+- `Key (LedgerState (ShelleyBlock proto era)) = TxIn (EraCrypto era)`
+- `Value (LedgerState (ShelleyBlock proto era)) = TxOut era`
+
+To instantiate the `mk` type variable, some _mapkinds_ are defined:
+
+| `MapKind :: Type -> Type -> Type` | Container                     | Used  for                                                         |
+|-----------------------------------|-------------------------------|------------------------------------------------------------------|
+| `ValuesMK k v`                    | `Map k v`                     | Ledger states passed to and from the Ledger rules                |
+| `KeysMK k v`                      | `Set k`                       | Querying the disk for the values needed by a block               |
+| `DiffMK k v`                      | `Map k (Delta v)`             | Carrying the differences created by applying a block             |
+| `EmptyMK k v`                     | $\emptyset$                   | When not needing info about the UTxO set, or the values are inside the `NewEpochState`                         |
+
+The actual invocation of the ledger rules make use of a `NewEpochState` which is unaware of any of this machinery. We use
+the `stowLedgerTables`/`unstowLedgerTables` functions to inject and project the values in the
+`NewEpochState` to the ledger tables, making this completely transparent for the Ledger layer.
+
+```haskell
+stowLedgerTables   :: l ValuesMK -> l EmptyMK
+unstowLedgerTables :: l EmptyMK  -> l ValuesMK
+```
+
+> ⚠️ It is very important to note that `EmptyMK` just means that _the ledger tables are empty_. This says nothing about whether there are values in the `NewEpochState`'s UTXO set. In the Consensus layer we take much care to ensure that the combination of `EmptyMK` having values in the internal UTXO set only happens at the Ledger layer boundary (via `stowLedgerTables`). Any other instance of `l EmptyMK` will mean that there are no values in the tables nor in the `NewEpochState`.
```

Comment:
It's not something I've considered much, though I'm not sure it would be worth it for the ledger to include it. It would be a pretty pervasive change, and it's a consensus implementation detail that is not so interesting from a

I think we currently have a *reasonably* safe API that prevent misuse. It could possibly be improved using types, e.g.:
* Introducing something like `StowedMK` / `UnstowedMK` instead of just `EmptyMK`
* Be more restrictive about when it is okay to extract a `NewEpochState` from a consensus `LedgerState`

Reply:
Probably this should be `StowedMK` / `EmptyMK` instead. `EmptyMK` meaning that the Ledger must have an empty UTxO map inside, `StowedMK` means that there are values in the UTxO set inside.

Reply:
#1334